### PR TITLE
BUG: special: fix array index out of bound for `pro_rad2`

### LINKED ISSUE #21461: BUG: `special.pro_rad2`: incorrect results since translation from Fortran
### Describe your issue.

In scipy 1.14.1, special function pro_rad2 returns incorrect values. In addition, multiple calls to pro_rad2 with the same input parameters will return differing results.

For `pro_rad2(0, 0, 3, 1.02)[0]`, 1.7434 is returned (or sometimes 3.4867, twice 1.7434). From tables in Flammer (https://github.com/scipy/scipy/issues/13968#issuecomment-831938400), the result should be -0.3422. 

I have Matlab code that looks to be using the same algorithms as scipy's C++ functions for pro_rad2() and this returns -0.3509, so similar to Flammer. I if try pro_rad2 in scipy v1.12, I get -0.3509, the same as the Matlab code. I understand that v1.12 is from before the translation of the spheroidal wave functions from Fortran to C++ (so perhaps a similar cause as #21058?).

The derivatives that pro_rad2() returns are also incorrect when compared to Flammer or the Matlab code.

### Reproducing Code Example

```python
from scipy.special import pro_rad2

for _ in range(30):
    print(pro_rad2(0, 0, 3, 1.02)[0])
```

### Error message

```shell
No error message generated.
```

### SciPy/NumPy/Python version and system information

```shell
1.14.1 1.26.4 sys.version_info(major=3, minor=11, micro=9, releaselevel='final', serial=0)
Build Dependencies:
  blas:
    detection method: pkgconfig
    found: true
    include directory: C:/Users/GavinMacaulay/miniforge3/envs/aqualyd/Library/include
    lib directory: C:/Users/GavinMacaulay/miniforge3/envs/aqualyd/Library/lib
    name: blas
    openblas configuration: unknown
    pc file directory: C:\bld\scipy-split_1724327218051\_h_env\Library\lib\pkgconfig
    version: 3.9.0
  lapack:
    detection method: pkgconfig
    found: true
    include directory: C:/Users/GavinMacaulay/miniforge3/envs/aqualyd/Library/include
    lib directory: C:/Users/GavinMacaulay/miniforge3/envs/aqualyd/Library/lib
    name: lapack
    openblas configuration: unknown
    pc file directory: C:\bld\scipy-split_1724327218051\_h_env\Library\lib\pkgconfig
    version: 3.9.0
  pybind11:
    detection method: pkgconfig
    include directory: C:/Users/GavinMacaulay/miniforge3/envs/aqualyd/Library/include
    name: pybind11
    version: 2.13.4
Compilers:
  c:
    commands: clang-cl
    linker: lld-link
    linker args: -Wl,-defaultlib:C:/bld/scipy-split_1724327218051/_build_env/lib/clang/19/lib/windows/clang_rt.builtins-x86_64.lib
    name: clang-cl
    version: 19.1.0-rc2
  c++:
    commands: clang-cl
    linker: lld-link
    linker args: -Wl,-defaultlib:C:/bld/scipy-split_1724327218051/_build_env/lib/clang/19/lib/windows/clang_rt.builtins-x86_64.lib
    name: clang-cl
    version: 19.1.0-rc2
  cython:
    commands: cython
    linker: cython
    name: cython
    version: 3.0.11
  fortran:
    args: -D_CRT_SECURE_NO_WARNINGS, -fms-runtime-lib=dll, -fuse-ld=lld, -IC:\bld\scipy-split_1724327218051\_h_env\Library\include
    commands: flang-new
    linker: lld-link
    linker args: -Wl,-defaultlib:C:/bld/scipy-split_1724327218051/_build_env/lib/clang/19/lib/windows/clang_rt.builtins-x86_64.lib,
      -D_CRT_SECURE_NO_WARNINGS, -fms-runtime-lib=dll, -fuse-ld=lld, -IC:\bld\scipy-split_1724327218051\_h_env\Library\include
    name: llvm-flang
    version: 19.1.0-rc2
  pythran:
    include directory: ..\..\_h_env\Lib\site-packages\pythran
    version: 0.16.1
Machine Information:
  build:
    cpu: x86_64
    endian: little
    family: x86_64
    system: windows
  cross-compiled: false
  host:
    cpu: x86_64
    endian: little
    family: x86_64
    system: windows
Python Information:
  path: C:\bld\scipy-split_1724327218051\_h_env\python.exe
  version: '3.11'
```

### PR DESCRIPTION
**Pull request by inkydragon**
#### Reference issue
Closes #21461

#### What does this implement/fix?
This pr fix some mistakes when translating f77 code into cpp.
- Array index out of bounds
	The array index of f77 sometimes starts at 0.
- Wrong variable type

### CODE REVIEW

File: scipy/special/xsf/specfun/specfun.h [hunk 1]
```diff
@@ -3582,10 +3582,10 @@ inline void kmn(int m, int n, T c, T cv, int kd, T *df, T *dn, T *ck1, T *ck2) {
     for (j = 1; j <= m; ++j)
         r5 = r5 * (j + m) / c;
 
-    g0 = dn[m - 1];
-
     if (m == 0)
         g0 = df[0];
+    else
+        g0 = dn[m - 1];
```
Comment (inkydragon) [inline scipy/special/xsf/specfun/specfun.h:3588]: With `m >= 0`, `dn[m - 1]` may result in index out-of-bounds.

File: scipy/special/xsf/specfun/specfun.h [hunk 2]
```diff
@@ -5024,14 +5024,16 @@ void rmn2sp(int m, int n, T c, T x, T cv, int kd, T *df, T *r2f, T *r2d) {
     //          and joining factors
     // ======================================================
 
-    int k, j, j1, j2, l1, ki, nm3, sum, sdm;
+    int k, j, j1, j2, l1, ki, nm3;
     T ip, nm1, nm, nm2, su0, sw, sd0, su1, sd1, sd2, ga, r1, r2, r3,\
-           sf, gb, spl, gc, sd, r4, spd1, spd2, su2, ck1, ck2;
+           sf, gb, spl, gc, sd, r4, spd1, spd2, su2, ck1, ck2, sum, sdm;
```
Comment (inkydragon) [inline scipy/special/xsf/specfun/specfun.h:5029]: With `IMPLICIT DOUBLE PRECISION (A-H,O-Z)`,
`sum, sdm` start with `s`, and `s ∈ [O-Z]`.

https://github.com/scipy/scipy/blob/9d08eb58205fb669ff72c60080b18be61b14d8e1/scipy/special/specfun/specfun.f#L585-L599

File: scipy/special/xsf/specfun/specfun.h [hunk 3]
```diff
@@ -5024,14 +5024,16 @@ void rmn2sp(int m, int n, T c, T x, T cv, int kd, T *df, T *r2f, T *r2d) {
...
 
+    // fortran index start from 0
     T *pm = (T *) malloc(252*sizeof(T));
     T *pd = (T *) malloc(252*sizeof(T));
     T *qm = (T *) malloc(252*sizeof(T));
     T *qd = (T *) malloc(252*sizeof(T));
+    // fortran index start from 1
     T *dn = (T *) malloc(201*sizeof(T));
```
Comment (inkydragon) [inline scipy/special/xsf/specfun/specfun.h:5037]: https://github.com/scipy/scipy/blob/9d08eb58205fb669ff72c60080b18be61b14d8e1/scipy/special/specfun/specfun.f#L600-L601

File: scipy/special/xsf/specfun/specfun.h [hunk 4]
```diff
@@ -5112,17 +5114,19 @@ void rmn2sp(int m, int n, T c, T x, T cv, int kd, T *df, T *r2f, T *r2d) {
     }
     su2 = 0.0;
     ki = (2 * m + 1 + ip) / 2;
+    ki = std::max(1, ki);
+    assert((ki-1) >= 0);
```
Comment (inkydragon) [inline scipy/special/xsf/specfun/specfun.h:5118]: I think this is a bug in f77.

With `pro_rad2(0, 0, 3, 1.02)`, we have `ki = 0, nm3 = 28`,
then `dn[k - 1]` may result in index out-of-bounds.

Using `max` to limit `ki` to a minimum of 1 is the easiest fix. The function also gives the correct result after the fix.
But I'm not sure if there is a better way, or if this has any effect on the precision.

 
https://github.com/scipy/scipy/blob/9d08eb58205fb669ff72c60080b18be61b14d8e1/scipy/special/specfun/specfun.f#L668-L674

**Review by inkydragon — COMMENTED** [review]
Add some notes to the changes.

**Comment by ilayn** [comment]
These look good to me. @steppi @izaid  could you also verify ?

**Comment by inkydragon** [comment]
@steppi  Maybe I need to open a new pr in https://github.com/scipy/xsf?

**Comment by tylerjereddy** [comment]
@steppi what would you like me to do here? We're planning to branch for `1.15.0` in a few days.

**Review by steppi — APPROVED** [review]
Thanks @inkydragon. Looks good. Let's get this in before we branch. I'll migrate the commit over to `xsf`.

**Pull request merged**